elastic/kibana
# [Infrastructure UI] Host filtering controls

Closes [#140445](https://github.com/elastic/kibana/issues/140445)

## Summary
This PR adds 2 filters (Operating System and Cloud Provider) using  [Kibana Controls API](https://github.com/elastic/kibana/tree/main/src/plugins/controls) to the Host view.

## Testing
- Open Host View
- The Operating System and Cloud Provider filters should be visible under the search bar. Supported values:
   - Filter Include/exclude OS name / Cloud Provider name
   - Exist / Does not exist
   - Any (also when clearing the filters) 
- The control filters should update the possible values when the other control filter or unified search query/filters are changes
- When the control group filters are updated the table is loading the filtered result.
- Combination with unified search query/filters should be possible.
![image](https://user-images.githubusercontent.com/14139027/203373557-f9220f22-53ee-4fe0-9bdd-cdc08ce31156.png)
- Copy the url after adding the filters and paste it into a separate tab 
   - The control group AND the other filters/query should be prefilled

## 🎉  UPDATE the control panels are prefilled from the URL 
### The Workaround:
Together with @ThomThomson we found a way to prefill the control group selections from the URL state by adding the panels' objects to the URL state (using a separate key to avoid the infinite loop issue) and keeping the output filters (used for updating the table) separately.

## Discovered issues with persisting the new filters to the URL state
~~⚠️ This PR does not support persisting those filters in the URL state. The reason behind this is that if we persist those filters inside the other filters parameter it will create an infinite loop (as those controls are relying on the filters to adjust the possible values).~~
In order to avoid that we can persist them in a different parameter (instead of adding them to the existing `_a` we can add a new one for example named `controlFilters`. This will work with filtering the table results. 
BUT If we go with the solution to persist them in another `urlStateKey` we also need to prefill those selections from the url state to the control filters (Operating System and Cloud Provide). 
Currently, the controls API supports setting `selectedOptions` as a string array. 
### Workaraound Ideas
Option 1: I tried first on a[ separate branch ](https://github.com/elastic/kibana/compare/main...jennypavlova:kibana:140445-host-filtering-controls-with-url-state) 
- Persist the filters as an array of filter options. 
- on load prefill the control filters 
   - extract the string values from the filters and set them as `selectedOptions` inside the control group input `panel` (based on the field name for example)

Option 2 (Suggestion from Devon ) 
- on load pass in the selections from the URL to the control group input
- Don't render the table right away
- Wait until control group is ready .then
   - Get the filters from the control group output
   - Set filters from controls in the use state by doing controls.getOutput().filters
- Render the table with ...unifiedSearchFilters, ...filtersFromControls

❌  The issue with both 1 & 2
- With `selectedOptions` we can prefill only **strings** so `Exist` and `Negate` won't be supported
